### DISCUSSION TITLE
does it work?

### DISCUSSION THREAD

**Discussion by lucidrains** [lucidrains/Adan-pytorch#1] [Announcements]
or is it another dud? let everyone know here

**Comment by lucidrains** [comment]
personally i saw a very slight improvement over adam on a toy autoregressive task, past 2k steps (slow start)

**Reply by lucidrains** [reply]
kept learning rate of adan at `1e-3` as `2e-4` was worse, then a sweep of adam from `1e-3` to `2e-4`. also tried the transformer-xl beta settings vs the other ones (which made less of a difference)

**Reply by JeanKaddour** [reply]
Can you please post plotted training curves? I am more interested in convergence speed than final performance.

**Comment by JeanKaddour** [comment]
Thank you so much for providing this implementation; especially so rapidly! 

I'm currently running this in a pipeline training a ResNet-34 on CIFAR100 with cosine scheduler (initial lr=0.1) for 200 epochs. It converges much more slower than SGD with momentum of 0.9 (the latter is AFAIK pretty much the default optimizer for these CNN-type tasks). 

I tried both `beta_1=0.02` (as reported in the paper) and `beta_1=0.9`; for the other betas, I stuck to their default value. Does this make sense? Anything I should improve without running a dedicated hparam search? In their paper, I can't find the beta values for such CNN-type tasks. 

I'm happy to post plotted training curves, but want to first check if I'm using this the (right) way I'm supposed to...

**Comment by hbredin** [comment]
Looks promising so far for training (permutation invariant) [neural speaker diarization](https://huggingface.co/pyannote/segmentation) 

Same learning rate (1e-3), default parameters for both Adam and Adan.

<img width="910" alt="image" src="https://user-images.githubusercontent.com/1080837/186984598-8332b1db-d4c0-403a-81c6-1a877b4cc7f4.png">

**Comment by Rocketknight1** [comment]
I made a TF port of the code that I think was valid, and got significantly slower convergence than just basic Adam on a ResNet-like model, but I didn't tune the LRs much.

**Reply by Rocketknight1** [reply]
That said, I should really verify correctness of the port before I say anything about performance, lol

**Comment by Rocketknight1** [comment]
Update: I confirmed correctness of my TF port (by running 20 epochs with it on the same toy model in PyTorch + TF and confirming that the weights were the same afterwards).

Performance is still **much** worse than Adam. I used my [minimal-lczero](https://github.com/Rocketknight1/minimal_lczero) repo and some of Leela's data for testing (ResNet-like architecture)

**Comment by rwightman** [comment]
So far only early convergence is standing out as clearly better than LAMB / AdamW with known good hparams. At this point hard to say if it's any differences in impl from paper or just requiring more exhaustive hparam search... later in training I'm usually ending up with the convergence slowing and ending up worse than comparalbe AdamW run... (for a hybrid ViT-CNN I'm fiddling with right now).

**Reply by rwightman** [reply]
It's a bit odd because the 'early convergence' improvement might suggest my LR is too low, but then the slow down through the middle phase might suggests it's too high... hmmm

**Reply by lucidrains** [reply]
thanks for weighing in Ross!

**Reply by JeanKaddour** [reply]
If you care about early convergence; maybe [this paper](https://arxiv.org/abs/2209.14981) is of interest to you

**Comment by conceptofmind** [comment]
Here are some tests for training a ViT-32 with adan vs adam on CIFAR10 for 100 epochs. I used the same learning rate of 1e-3 for both runs. I left the other hyperparameters as defaults. The models and training scripts are identical except for changing the optimizer.

Adan training results:

Train/Validation Loss
![Screenshot from 2022-11-12 11-16-13](https://user-images.githubusercontent.com/25208228/201483645-268ce565-fde9-4040-ac76-493a66108b55.png)

Train/Validation Accuracy
![Screenshot from 2022-11-12 11-15-11](https://user-images.githubusercontent.com/25208228/201483616-f1801839-661f-48fa-bdbd-07efef1cba24.png)

Adam training results:

Train/Validation Loss:
![Screenshot from 2022-11-12 11-17-50](https://user-images.githubusercontent.com/25208228/201483724-224afebd-5eef-4fbf-a794-1eb6246e15c1.png)

Train/Validation Accuracy:
![Screenshot from 2022-11-12 11-18-38](https://user-images.githubusercontent.com/25208228/201483781-b9a08962-6494-4399-97b5-61191b04fa6c.png)

I will have to do another 2 or 4 runs and average the results. Adan looks promising from this initial test.

Best,

Enrico
